### PR TITLE
update actions/(download|upload)-artifact

### PR DESCRIPTION
**Pull request by dholth**
### Description

Update actions to latest versions to avoid "node deprecated" warnings

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Review by dholth — COMMENTED** [review]
https://github.com/actions/upload-artifact/commit/65462800fd760344b1a7b4382951275a0abb4808

https://github.com/actions/download-artifact/commit/65a9edc5881444af0b9093a5e628f2fe47ea3b2e

File: .github/workflows/canary-weekly.yml [hunk 1]
```diff
@@ -68,12 +68,12 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Download test results
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@65a9edc5881444af0b9093a5e628f2fe47ea3b2e
```
Comment (dholth) [inline .github/workflows/canary-weekly.yml:71]:
```suggestion
        uses: actions/download-artifact@65a9edc5881444af0b9093a5e628f2fe47ea3b2e # v4.1.7
```

File: .github/workflows/canary-weekly.yml [hunk 2]
```diff
@@ -68,12 +68,12 @@ jobs:
...
 
       - name: Upload combined test results
         # provides one downloadable archive of all .coverage/test-report.xml files
         # of all matrix runs for further analysis.
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@65462800fd760344b1a7b4382951275a0abb4808
```
Comment (dholth) [inline .github/workflows/canary-weekly.yml:76]:
```suggestion
        uses: actions/upload-artifact@65462800fd760344b1a7b4382951275a0abb4808 # v4.3.3
```

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -61,12 +61,12 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Download test results
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@65a9edc5881444af0b9093a5e628f2fe47ea3b2e
```
Comment (dholth) [inline .github/workflows/tests.yml:64]:
```suggestion
        uses: actions/download-artifact@65a9edc5881444af0b9093a5e628f2fe47ea3b2e # v4.1.7
```

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -61,12 +61,12 @@ jobs:
...
 
       - name: Upload combined test results
         # provides one downloadable archive of all .coverage/test-report.xml files
         # of all matrix runs for further analysis.
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@65462800fd760344b1a7b4382951275a0abb4808
```
Comment (dholth) [inline .github/workflows/tests.yml:69]:
```suggestion
        uses: actions/upload-artifact@65462800fd760344b1a7b4382951275a0abb4808 # v4.3.3
```

**Review by dbast — APPROVED** [review]

**Pull request merged**